### DISCUSSION TITLE
How to add an additional variable type?

### DISCUSSION THREAD

**Discussion by Ruslan618** [microsoft/vscode-cpptools#9737] [Q&A] [answered]
I am a MQL programmer and I really like VS Code, but there is no official support for it. For this I use the C/C++ IntelliSense extension through the .cpp file association.  MQL is similar to C++. Thus, most of the functionality works fine, but unfortunately not all. For example, C/C++ IntelliSense does not see such a variable type as "string" (it is used in MQL anyway) and all the advantages of IntelliSense for such variables, functions and methods do not work.
![2022-08-13_12-54-28](https://user-images.githubusercontent.com/88244331/184478954-b1d6142d-66fa-4670-b521-06dd83549cc3.png)
Here is an example - the screenshot shows that the string variable is not available. Hence the question, is it possible somehow for C / C ++ IntelliSense to add an additional type of variable "string" so that it is fully processed? Is it in principle possible?
I really want to fully use VS Code, so thanks in advance for any help.

**Comment by sean-mcmanus** [comment, marked as answer]
You could set the "forcedInclude" property to be some file with additional definitions or use some code like
```cpp
#ifdef __INTELLISENSE__
// other defs
#endif
```
in an existing header.

Also, I'm assuming you're using the "C/C++" extension (this extension). If you're using the "C/C++ Intellisense" extension that you would be a different one.

**Reply by Ruslan618** [reply]
Thank you for your reply, but could you be answer more details? Yes, I use this extension but I have no idea how to do it. May be with configurate file compile_commands.json?

**Reply by sean-mcmanus** [reply]
You're asking how to set the forcedInclude? Yeah, that depends on how you're configuring, i.e. if you're using c_cpp_properties.json directly or setting compileCommands or configurationProvider. You can run C/C++: Log Diagnostics to check if the forcedInclude setting is actually being used or not in the current file (if it weren't use, you'd also see the "string" as an unrecognized type as well).

**Reply by Ruslan618** [reply]
Ok tnank you, i will try.

**Reply by Ruslan618** [reply]
Until I got nothing. I use c_cpp_properties.json, can you plz detail describe what me need to do to add type var of string?

**Reply by sean-mcmanus** [reply]
You could add something like `using string = const char *;` to your forced include file. Or if the string type has other methods you could add a class that defines what it supports.

**Reply by Ruslan618** [reply]
Tnank you, but my compiler MQL dont admit this. I was thinking to add a variable type at the extension level, or in VS code settings, that this worked for all files. Maybe it's still possible?!

**Reply by sean-mcmanus** [reply]
Does MQL have a typedef or equivalent?

The code would be in a `#ifdef __INTELLISENSE__` block so your MQL shouldn't see the code. What is the issue you're having?

**Reply by Ruslan618** [reply]
Yes MQL can use #ifdef __INTELLISENSE__ and I managed to use the construction using "string = const char *;" I don’t understand why I didn’t succeed right away and how it works). Now the string variables work correctly. Thank you very much, you helped me a lot.

**Reply by Ruslan618** [reply]
It's only interesting now that when you enter a string variable, it is indicated by a different icon, unlike ordinary variables.
![2022-09-24_11-04-08](https://user-images.githubusercontent.com/88244331/192088009-46f94b1c-38cc-4d8b-a205-4eee190a0cf0.png)
![2022-09-24_11-05-40](https://user-images.githubusercontent.com/88244331/192087931-fc9b49ba-220e-49e2-8de2-1758c58dc4ce.png)
Otherwise, everything works as it should.

**Reply by sean-mcmanus** [reply]
The loop icon is for a typedef.  The abc one is just text and the other one is for keywords.
